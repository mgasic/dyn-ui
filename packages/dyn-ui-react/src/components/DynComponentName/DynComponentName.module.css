.root {
  display: var(--dyn-component-name-display, block);
  box-sizing: border-box;
  padding: var(--dyn-component-name-padding, 0);
  margin: var(--dyn-component-name-margin, 0);
  gap: var(--dyn-component-name-gap, 0);
}

@media (min-width: 40rem) {
  .root {
    padding: var(
      --dyn-component-name-padding-sm,
      var(--dyn-component-name-padding, 0)
    );
    margin: var(
      --dyn-component-name-margin-sm,
      var(--dyn-component-name-margin, 0)
    );
    gap: var(
      --dyn-component-name-gap-sm,
      var(--dyn-component-name-gap, 0)
    );
  }
}

@media (min-width: 48rem) {
  .root {
    padding: var(
      --dyn-component-name-padding-md,
      var(--dyn-component-name-padding-sm, var(--dyn-component-name-padding, 0))
    );
    margin: var(
      --dyn-component-name-margin-md,
      var(--dyn-component-name-margin-sm, var(--dyn-component-name-margin, 0))
    );
    gap: var(
      --dyn-component-name-gap-md,
      var(--dyn-component-name-gap-sm, var(--dyn-component-name-gap, 0))
    );
  }
}

@media (min-width: 64rem) {
  .root {
    padding: var(
      --dyn-component-name-padding-lg,
      var(--dyn-component-name-padding-md, var(--dyn-component-name-padding, 0))
    );
    margin: var(
      --dyn-component-name-margin-lg,
      var(--dyn-component-name-margin-md, var(--dyn-component-name-margin, 0))
    );
    gap: var(
      --dyn-component-name-gap-lg,
      var(--dyn-component-name-gap-md, var(--dyn-component-name-gap, 0))
    );
  }
}

@media (min-width: 80rem) {
  .root {
    padding: var(
      --dyn-component-name-padding-xl,
      var(--dyn-component-name-padding-lg, var(--dyn-component-name-padding, 0))
    );
    margin: var(
      --dyn-component-name-margin-xl,
      var(--dyn-component-name-margin-lg, var(--dyn-component-name-margin, 0))
    );
    gap: var(
      --dyn-component-name-gap-xl,
      var(--dyn-component-name-gap-lg, var(--dyn-component-name-gap, 0))
    );
  }
}
