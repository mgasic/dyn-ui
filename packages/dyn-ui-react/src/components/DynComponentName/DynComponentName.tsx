import React, { forwardRef, useMemo } from 'react';
import { cn } from '../../utils/classNames';
import type {
  DynComponentNameBreakpoint,
  DynComponentNameProps,
  DynComponentNameRef,
  DynComponentNameResponsiveSpacingValue,
  DynComponentNameSpacingToken,
} from './DynComponentName.types';
import styles from './DynComponentName.module.css';

const RESPONSIVE_BREAKPOINTS: Exclude<DynComponentNameBreakpoint, 'base'>[] = ['sm', 'md', 'lg', 'xl'];

const SPACING_TOKENS: Record<Exclude<DynComponentNameSpacingToken, '0' | 'auto'>, string> = {
  xs: '0.25rem',
  sm: '0.5rem',
  md: '1rem',
  lg: '1.5rem',
  xl: '2rem',
  '2xl': '3rem',
};

const isResponsiveObject = (
  value: DynComponentNameResponsiveSpacingValue | undefined
): value is Partial<Record<DynComponentNameBreakpoint, DynComponentNameResponsiveSpacingValue>> =>
  typeof value === 'object' && value !== null;

const resolveSpacingValue = (
  value: DynComponentNameResponsiveSpacingValue | undefined
): {
  base?: DynComponentNameResponsiveSpacingValue;
  responsive: Partial<Record<Exclude<DynComponentNameBreakpoint, 'base'>, DynComponentNameResponsiveSpacingValue>>;
} => {
  if (value === undefined) {
    return { responsive: {} };
  }

  if (!isResponsiveObject(value)) {
    return { base: value, responsive: {} };
  }

  const responsive: Partial<Record<Exclude<DynComponentNameBreakpoint, 'base'>, DynComponentNameResponsiveSpacingValue>> = {};

  for (const breakpoint of RESPONSIVE_BREAKPOINTS) {
    const token = value[breakpoint];
    if (token !== undefined) {
      responsive[breakpoint] = token;
    }
  }

  return {
    base: value.base,
    responsive,
  };
};

type SpacingSlot = 'padding' | 'margin' | 'gap';

const toSpacingCssValue = (
  value: DynComponentNameResponsiveSpacingValue,
  slot: SpacingSlot
): string => {
  if (typeof value === 'number') {
    return `${value}px`;
  }

  if (value === 'auto') {
    return slot === 'margin' ? 'auto' : String(value);
  }

  if (value === '0') {
    return '0';
  }

  const token = SPACING_TOKENS[value as Exclude<DynComponentNameSpacingToken, '0' | 'auto'>];
  if (token) {
    return `var(--dyn-spacing-${value}, var(--spacing-${value}, ${token}))`;
  }

  return String(value);
};

const assignSpacingVars = (
  slot: SpacingSlot,
  cssValue: string,
  target: React.CSSProperties,
  breakpoint?: Exclude<DynComponentNameBreakpoint, 'base'>
) => {
  const suffix = breakpoint ? `-${breakpoint}` : '';
  target[`--dyn-component-name-${slot}${suffix}`] = cssValue;
};

const buildSpacingVars = (
  gap: DynComponentNameResponsiveSpacingValue | undefined,
  padding: DynComponentNameResponsiveSpacingValue | undefined,
  margin: DynComponentNameResponsiveSpacingValue | undefined
) => {
  const vars: React.CSSProperties = {};

  const applySpacing = (value: DynComponentNameResponsiveSpacingValue | undefined, slot: SpacingSlot) => {
    if (value === undefined) {
      return;
    }

    const { base, responsive } = resolveSpacingValue(value);

    if (base !== undefined) {
      assignSpacingVars(slot, toSpacingCssValue(base, slot), vars);
    }

    for (const [breakpoint, token] of Object.entries(responsive)) {
      if (token !== undefined) {
        assignSpacingVars(
          slot,
          toSpacingCssValue(token, slot),
          vars,
          breakpoint as Exclude<DynComponentNameBreakpoint, 'base'>
        );
      }
    }
  };

  applySpacing(gap, 'gap');
  applySpacing(padding, 'padding');
  applySpacing(margin, 'margin');

  return vars;
};

export const DynComponentName = forwardRef(function DynComponentNameInner<
  E extends React.ElementType = 'div'
>(
  props: DynComponentNameProps<E>,
  ref: DynComponentNameRef<E>
) {
  const {
    as,
    gap,
    p,
    m,
    className,
    style,
    children,
    'data-testid': dataTestId = 'dyn-component-name',
    ...rest
  } = props;

  const Component = (as ?? 'div') as React.ElementType;

  const spacingVars = useMemo(() => buildSpacingVars(gap, p, m), [gap, p, m]);

  return (
    <Component
      {...rest}
      ref={ref}
      data-testid={dataTestId}
      className={cn(styles.root, className)}
      style={{ ...spacingVars, ...style }}
    >
      {children}
    </Component>
  );
});

DynComponentName.displayName = 'DynComponentName';
