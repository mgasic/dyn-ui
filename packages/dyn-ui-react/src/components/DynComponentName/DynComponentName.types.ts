import type { ComponentPropsWithoutRef, ComponentRef, ElementType, ReactNode } from 'react';

export type DynComponentNameElement = ElementType;

export type DynComponentNameSpacingToken = '0' | 'xs' | 'sm' | 'md' | 'lg' | 'xl' | '2xl' | 'auto';

export type DynComponentNameBreakpoint = 'base' | 'sm' | 'md' | 'lg' | 'xl';

export type DynComponentNameSpacingValue = DynComponentNameSpacingToken | string | number;

export type DynComponentNameResponsiveSpacingValue =
  | DynComponentNameSpacingValue
  | Partial<Record<DynComponentNameBreakpoint, DynComponentNameSpacingValue>>;

export type DynComponentNameOwnProps = {
  /**
   * Change the underlying HTML element while keeping all spacing behaviour intact.
   * Accepts both native tags and custom React components.
   */
  as?: DynComponentNameElement;
  /** Optional gap spacing token or raw CSS length. */
  gap?: DynComponentNameResponsiveSpacingValue;
  /** Optional padding spacing token or raw CSS length. */
  p?: DynComponentNameResponsiveSpacingValue;
  /** Optional margin spacing token or raw CSS length. */
  m?: DynComponentNameResponsiveSpacingValue;
  /** Custom className merged with the component root. */
  className?: string;
  /** Inline style overrides merged after computed CSS variables. */
  style?: ComponentPropsWithoutRef<'div'>['style'];
  /** Test id for querying in tests. */
  'data-testid'?: string;
  children?: ReactNode;
};

export type DynComponentNameProps<E extends ElementType = 'div'> = DynComponentNameOwnProps &
  Omit<ComponentPropsWithoutRef<E>, keyof DynComponentNameOwnProps>;

export type DynComponentNameRef<E extends ElementType = 'div'> = ComponentRef<E>;
