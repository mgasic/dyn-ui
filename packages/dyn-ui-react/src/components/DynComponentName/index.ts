export { DynComponentName } from './DynComponentName';
export type {
  DynComponentNameProps,
  DynComponentNameRef,
  DynComponentNameOwnProps,
  DynComponentNameElement,
  DynComponentNameSpacingToken,
  DynComponentNameBreakpoint,
  DynComponentNameSpacingValue,
  DynComponentNameResponsiveSpacingValue,
} from './DynComponentName.types';
