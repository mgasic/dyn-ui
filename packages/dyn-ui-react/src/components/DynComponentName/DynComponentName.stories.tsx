import type { Meta, StoryObj } from '@storybook/react';
import { DynComponentName } from './DynComponentName';

const meta: Meta<typeof DynComponentName> = {
  title: 'Layout/DynComponentName',
  component: DynComponentName,
  args: {
    p: 'md',
    gap: 'sm',
    children: (
      <>
        <span>Primary content</span>
        <span>Secondary content</span>
      </>
    ),
    style: { display: 'flex', flexDirection: 'column' },
  },
};

export default meta;

type Story = StoryObj<typeof DynComponentName>;

export const Default: Story = {};

export const AsSection: Story = {
  args: {
    as: 'section',
    m: 'lg',
    children: (
      <>
        <h2 style={{ margin: 0 }}>Section heading</h2>
        <p style={{ margin: 0 }}>Content rendered inside a semantic section element.</p>
      </>
    ),
  },
};

export const AsList: Story = {
  args: {
    as: 'ul',
    gap: 'xs',
    p: 'sm',
    style: { listStyle: 'disc', display: 'flex', flexDirection: 'column' },
    children: (
      <>
        <li>Item one</li>
        <li>Item two</li>
        <li>Item three</li>
      </>
    ),
  },
};

export const ResponsiveSpacing: Story = {
  args: {
    as: 'article',
    p: { base: 'sm', md: 'lg' },
    m: { base: '0', lg: 'xl' },
    gap: { base: 'xs', lg: 'md' },
    children: (
      <>
        <h3 style={{ margin: 0 }}>Responsive spacing</h3>
        <p style={{ margin: 0 }}>
          Padding, margin and gap respond to viewport width using Dyn UI spacing tokens.
        </p>
      </>
    ),
  },
};
