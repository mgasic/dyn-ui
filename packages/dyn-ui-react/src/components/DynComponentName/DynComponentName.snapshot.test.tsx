import { render } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import { DynComponentName } from './DynComponentName';

describe('DynComponentName', () => {
  it('matches snapshot with default settings', () => {
    const { container } = render(
      <DynComponentName p="md" gap="sm">
        <span>Snapshot content</span>
        <span>Secondary line</span>
      </DynComponentName>
    );

    expect(container).toMatchSnapshot();
  });

  it('supports polymorphic rendering and responsive spacing', () => {
    const { container } = render(
      <DynComponentName
        as="section"
        p={{ base: 'sm', lg: 'xl' }}
        m={{ base: '0', md: 'lg' }}
        gap={{ base: 'xs', md: 'lg' }}
      >
        <h2>Section title</h2>
        <p>Supporting copy for the section.</p>
      </DynComponentName>
    );

    expect(container).toMatchSnapshot();
  });
});
