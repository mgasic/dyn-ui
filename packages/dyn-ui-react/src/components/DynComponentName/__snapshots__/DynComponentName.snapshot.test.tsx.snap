// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DynComponentName > matches snapshot with default settings 1`] = `
<div>
  <div
    class="_root_754333"
    data-testid="dyn-component-name"
    style="--dyn-component-name-gap: var(--dyn-spacing-sm, var(--spacing-sm, 0.5rem)); --dyn-component-name-padding: var(--dyn-spacing-md, var(--spacing-md, 1rem));"
  >
    <span>
      Snapshot content
    </span>
    <span>
      Secondary line
    </span>
  </div>
</div>
`;

exports[`DynComponentName > supports polymorphic rendering and responsive spacing 1`] = `
<div>
  <section
    class="_root_754333"
    data-testid="dyn-component-name"
    style="--dyn-component-name-gap: var(--dyn-spacing-xs, var(--spacing-xs, 0.25rem)); --dyn-component-name-gap-md: var(--dyn-spacing-lg, var(--spacing-lg, 1.5rem)); --dyn-component-name-padding: var(--dyn-spacing-sm, var(--spacing-sm, 0.5rem)); --dyn-component-name-padding-lg: var(--dyn-spacing-xl, var(--spacing-xl, 2rem)); --dyn-component-name-margin: 0; --dyn-component-name-margin-md: var(--dyn-spacing-lg, var(--spacing-lg, 1.5rem));"
  >
    <h2>
      Section title
    </h2>
    <p>
      Supporting copy for the section.
    </p>
  </section>
</div>
`;
